rfcs/unboxed-types.md

```diff
@@ -0,0 +1,458 @@
+have some subset of the following.
+
+**Variants**
+
+So far, we've only discussed records, when unboxed variants are
+equally useful. The general case is quite tricky, but several special
+cases are easy and useful: variants containing only constant
+constructors, or variants whose non-constant constructors have the
+same types.
+
+The tricky part of the general case is that the layout of a variant
+(including number of fields and whether those fields need to be
+registered with the GC) varies according to the tag. One simple way
+around this difficulty is to always allocate enough space for all
+fields: this might make sense for small variants like `result`, where
+the savings from unboxing outweigh the extra word used to represent them.
+
+**Nullable types**
+
+Unboxing the `'a option` variant would improve efficiency, but an
+unboxed `t option` still requires two words: one to store the `t` (if
+`Some`), and one to store the tag (`Some` or `None`). Picking a
+special magic value for `None` isn't enough to cram options into one
+word, since we need to be able to distinguish `None` from `Some None`,
+`Some (Some None)`, etc.
+
+In many cases, we know statically that the argument to option cannot
+be another option (e.g. `string option`, `int option`, `float
+option`...). In these cases, we can use *nullable types* instead. In a
+nullable type, the type parameter `'a` has layout `non_nullable`,
+which is a subtype of `value`. The `Some` constructor for nullable
+types is the identity (and in particular, does not allocate), while
+the `None` constructor is a constant null value.
+
+Since the equivalent of `t option option` gives a type error, no
+ambiguity is introduced by implementing `Some` as the identity.
+
+**Padding**
+
+As well as being able to control boxing of types, it's sometimes also
+useful to control the bytes of padding between unboxed elements. (This
+is sometimes useful for optimisation, but the main application is
+being able to directly map externally-specified formats like protocols
+or C structs).
+
+**Bitpacking**
+
+Related to byte-level padding, but quite different in implementation,
+bitpacking allows multiple unboxed fields (say, boolean flags) to
+occupy the same byte. In particular, a record containing <32 booleans
+could be given an `int32` layout rather than a product layout (and
+thus be passed around in just a single register).
+
+**GADTs**
+
+As well as refining types, GADTs should be able to refine layouts as
+well. It would be nice to be able to define the following:
+
+    type _ is_immediate =
+    | Immediate : ('a : immediate) -> 'a is_immediate
+    | Other : ('a : value) -> 'a is_immediate
+
+
+## Future projects
+
+**Monomorphisation**
+
+"Uncompilable" types like `diag : ('a : any_value) . 'a -> 'a * 'a`
+can actually be compiled, by compiling a separate version of `diag`
+for each layout. In fact, we need fewer layouts than that: we can
+share the same version for `value` and `immediate`. Generally,
```

Comment:
@gasche says
Note: another possible interpretation of any_value (than specialization) would be similar to F#'s "statically resolved type parameters", where the declaration does not result in any code being generated, but it has to be inlined at each callsite (it's an error if it's used in a non-inlinable way somewhere).

That said, I'm not personally fond of either proposals, as I think that both specialization and static-resolution are somewhat orthogonal to sorts/layout; we may want to specialize a definition according to the specialization abilities of polymorphic comparison operators (int, string, float...), and to statically resolve something to statically determine a modular-implicit parameter, rather than a type.

 @lpw25 says
I think that F#'s thing is pretty much a less expressive version of the "Macros" section below. It's probably not clear from the description, but what that section has in mind is the macros work described in the "modular macros" talk and here. With that you can allow sort polymorphic macros producing quotations. Splicing these in at a particular sort is basically the same as calling a function with "statically resolved type parameters" in F#.

Of course, as the below section says, this stuff should really be left for a later project. I think Stephen only included it here because if he didn't people would only ask about it anyway.